src/Microsoft.AspNetCore.ResponseCaching/ResponseOutput/ResponseCacheStream.cs

```diff
@@ -12,37 +12,56 @@ internal class ResponseCacheStream : Stream
     {
         private readonly Stream _innerStream;
         private readonly long _maxBufferSize;
+        private readonly int _shardSize;
+        private WriteOnlyShardStream _writeOnlyStream;
+        private ReadOnlyShardStream _readOnlyStream;
 
-        public ResponseCacheStream(Stream innerStream, long maxBufferSize)
+        internal ResponseCacheStream(Stream innerStream, long maxBufferSize, int shardSize)
         {
             _innerStream = innerStream;
             _maxBufferSize = maxBufferSize;
+            _shardSize = shardSize;
+            _writeOnlyStream = new WriteOnlyShardStream(_shardSize);
         }
 
-        public MemoryStream BufferedStream { get; } = new MemoryStream();
-
-        public bool BufferingEnabled { get; set; } = true;
+        internal bool BufferingEnabled { get; private set; } = true;
 
         public override bool CanRead => _innerStream.CanRead;
 
         public override bool CanSeek => _innerStream.CanSeek;
 
         public override bool CanWrite => _innerStream.CanWrite;
 
-        public override long Length => _innerStream.Length;
+        public override long Length => _writeOnlyStream.Length;
```

Comment:
Except ResponseCacheStream is exposed directly to the application. If the underlying stream is tracking length then you'll return the incorrect answer in scenarios where buffering is disabled.